### ISSUE TITLE
Content doesn't scroll in Edge

### ISSUE THREAD

**Issue by k-kolev1985** [nvaccess/nvda#7075]
In Microsoft Edge, when NVDA is reading simple text (with no focusable objects like links, buttons, etc. inside it), the web page does not scroll to show the currently read text. So, if the user reads a long segment of simple text from top to bottom, Edge will show only the part which currently fits vertically on the screen, no matter that NVDA has passed reading that part and it is reading way down the text. If NVDA encounters a focusable element on the page (e.g. a button, a link, etc.), Edge will scroll the view to show that focusable element/object on the screen. Is this a limitation in Edge or a bug in NVDA?

Steps to reproduce it:
1. Start NVDA.
2. Open Microsoft Edge and load a page in it, which contains a long segment (it may be of separate paragraphs) of simple text (with no focusable elements/objects like links, buttons, etc. in it). A good example is the following: [https://www.nvaccess.org/about/our-story](url)
3. Start reading the "Our Story" article with the arrow keys.

Expected results:
NVDA should scroll the web page, so it will constantly show on the screen what it is currently reading.

Actual results:
The web page does not scroll.

Test environment:
- Operating system: Windows 10 Pro (build 15063.138), 64-bit, in Bulgarian with all locale settings set to "Bulgarian".
- NVDA version: next-13994,744a2d39.
- Microsoft Edge version: 40.15063.0.0.
- Microsoft EdgeHTML version: 15.15063.
- Processor: Intel Core i5-2320 at 3.00GHz.
- RAM Memory: 8.00GB.
- Graphics: Palit GeForce GT 630, 1024MB dedicated memory + 3071MB shared system memory, desktop resolution set to 1920x1080.
- Sound Card: Realtek ALC662 at Intel Cougar Point PCH - High Definition Audio Controller.

**Comment by Brian1Gaff** [comment]
I'm almost certain this also happens in Internet Exploder and has since I 
can remember as I've often done a say all and a sighted companion has stated 
that the text I am hearing is not on the screen and no scrolling has 
happened. Myself, when alone I cannot course check this.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal email to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message -----

**Comment by k-kolev1985** [comment]
Actually, when using "Say All", this occurs in Firefox and Internet Explorer as well. And that is an entirely different bug, if I'm not mistaking. And that one has been filed here, also if I'm not mistaking.

If we read manually by pressing DOWN ARROW however, this is OK for Firefox and Internet Explorer, but it is broken for Edge.

NOTE: I'm mentioning only those 3 browsers, because I don't have experience with other supported by NVDA browsers, like Chrome for example.

**Comment by Brian1Gaff** [comment]
I tend not to read line by line if I'm reading articles, so yes, I agree it 
does also happen in Firefox and I suspect Chrome from what I've heard since 
this thread started here. Sorry for muddying the waters.
 Brian

**Comment by jcsteh** [comment]
CC @MichaelDCurran. We need to implement scrollIntoView somehow. Does Edge support the ScrollItem pattern for all elements? Alternatively, I see TextRange has a ScrollIntoView. I wonder if that's implemented? That doesn't quite map to how we do it elsewhere (text instead of object), but it theoretically gives better granularity and we can probably tweak things to fit it somehow.

**Comment by leonardder** [comment]
@k-kolev1985: Could you please try [this](https://ci.appveyor.com/api/buildjobs/qp13ncs9yc3p2biy/artifacts/output%2Fnvda_snapshot_try-UIAScroll-16915%2C0a7d239e.exe) try build?

**Comment by k-kolev1985** [comment]
@leonardder I've done some small test on a few pages which contain a lot of paragraphs containing only text, in which cases by default the scrolling is problematic. With this test build of NVDA, the scrolling seams to be handled fine.

I've noticed a few things though:
1. The scrolling is based on the bottom edge of the screen, no matter if we are reading down or up the page. Is it a limitation or by design?
2. When reading down the page by paragraphs (Control+Down arrow key), scrolling brings into view only the first line of the paragraph, and not the whole paragraph.

**Comment by leonardder** [comment]
[**@k-kolev1985**](https://github.com/k-kolev1985) commented on [14 Mar 2019, 17:14 CET](https://github.com/nvaccess/nvda/issues/7075#issuecomment-472941702 "2019-03-14T16:14:39Z - Replied by Github Reply Comments"):
> 1. The scrolling is based on the bottom edge of the screen, no matter if we are reading down or up the page. Is it a limitation or by design?

With UIA, it is required to provide a preference for either the bottom or the top of the page. This is also possible for IAccessible2, so I guess I can look into a method that checks whether the new content should be shown either at the top or bottom of the screen.

> 2. When reading down the page by paragraphs (Control+Down arrow key), scrolling brings into view only the first line of the paragraph, and not the whole paragraph.

That's because the caret is at the start of the paragraph, so the caret itself isn't taken into account. I guess that's a limitation we'll have to deal with, unless someone else has an idea how to fix this.

**Comment by leonardder** [comment]
Could you elaborate on whether te text scrolls on every key press, or only when the browse mode caret would otherwise go out of screen?

**Comment by k-kolev1985** [comment]
When reading down the page, scrolling occurs when the virtual cursor gets to the last visible line on the page. To be more accurate, the currently read line is always on the bottom of the page, whenever applicable.

Scrolling occurs on every key press it seams, because I think NVDA reads the text line per line at each key press.

**Comment by leonardder** [comment]
So just to make sure.

- If moving downwards with down arrow, the read line is always the bottom line on the screen?
- When moving upwards with up arrow, the current line will also be the bottom line of the screen?

**Comment by k-kolev1985** [comment]
Yes, exactly. Exceptions are the cases where the page has reached its top or bottom and cannot be scrolled further.

**Comment by leonardder** [comment]
Ugh, that's not very helpful. I'll try to code another, actually much simpler, approach later today.
Technically, it turns out that scrollIntoView on an UIA text range always scrolls either to the top or bottom of the page. There's no way to say, only scroll when the text is currently out of the view, so we'd have to find that out manually. I guess I'll revert this to calling scrollIntoView on the ScrollItem pattern as originally suggested by @jcsteh.

**Comment by k-kolev1985** [comment]
Yes, now I understood exactly what you meant by "always scrolling". Even if I scroll the text/page with the mouse so the currently read line is somewhere in the middle of the screen for example, as soon as I continue reading with the up/down arrow keys, the view scrolls so the current line is on the bottom.

**Comment by leonardder** [comment]
[Here](https://ci.appveyor.com/api/buildjobs/05wk0yt4xnvbmlt0/artifacts/output%2Fnvda_snapshot_try-UIAScroll-16918%2C4a736e29.exe) is a new try build.

**Comment by k-kolev1985** [comment]
Unfortunately, this behaves worse. At some point, even though there is no need for it yet, it scrolls the page so the currently read line is at the top. But after that, it does not scroll until it gets to a focusable object (e.g. a button). [This page](https://www.fanfiction.net/s/10288529/22/Visitation) is a good example for the behavior.

**Comment by leonardder** [comment]
I finally picked this up again.

> Technically, it turns out that scrollIntoView on an UIA text range always scrolls either to the top or bottom of the page. There's no way to say, only scroll when the text is currently out of the view, so we'd have to find that out manually.

So that's what I've done now:

@k-kolev1985 : Could you please test [this](https://ci.appveyor.com/api/buildjobs/iak0fpesqhjj176a/artifacts/output%2Fnvda_snapshot_try-UIAScroll-17065%2Cf62d796a.exe) try build?

**Comment by k-kolev1985** [comment]
OK, this is much better. Now the scrolling is as follows:
1. The scrolling when reading down the page occurs when the virtual cursor reaches the bottom visible part of the page. Pressing down arrow reads line by line, so scrolling occurs line by line.
2. The scrolling when reading up the page occurs when the virtual cursor reaches the top visible part of the page. Pressing up arrow reads line by line, so scrolling occurs line by line.
For me, this seams OK.

**Comment by leonardder** [comment]
This is exactly how I intended things to be, so that's great!

**Comment by leonardder** [comment]
I'm taking the scrollIntoView issues reported in #6382 as part of this as well.

@Qchristensen wrote in https://github.com/nvaccess/nvda/issues/6382#issuecomment-316315798:

> Re how to scroll when reading, I would suggest if something is on screen, don't scroll, only scroll if moving to something (either the next line down or the next element of a particular kind) is off screen. In that case, if the behaviour was to scroll the element to the top of the screen, then pressing down arrow to move through text would essentially move through one screen, then scroll one screen, then move through that screen, then scroll, which should work (from a user perspective) quite well. If the screen was only scrolled to bring the element to the bottom line of the screen, then this would be in line with what happens in say, Word or a text editor when down arrowing through text.

I'm inclined to implement the latter as in the try builds above, but both are possible. It would help if either @Qchristensen or @feerrenrut can give an opinion on this. Also @k-kolev1985, what do you prefer?

**Comment by leonardder** [comment]
I created [a new try build](https://ci.appveyor.com/api/buildjobs/xptxwh1fluqtl7p6/artifacts/output%2Fnvda_snapshot_try-UIAScroll-17079%2Ce2ee4106.exe) that should give identical results across the Chrome, Edge and FIrefox browsers. @k-kolev1985: Would you be able to confirm this?

**Comment by k-kolev1985** [comment]
I like the current implementation in the "try" builds.

I've noticed however, that the behavior is a bit different between Edge and Firefox (I don't use Chrome). In Edge, when the cursor has reached the bottom line and we've read a few more lines down and the page has scrolled, if we press up arrow to read a line up, scrolling is not done, because it is not necessary. In Firefox however, if we do the same and read a line up, the page is scrolled so the line to be read is at the top of the view and the scrolling up is done if we continue reading up and the read line is on top of the view. Hope I've managed to explain it correctly.

P.S.: Scrolling with "say all" does not work, but I guess that is a separate issue (?).

**Comment by leonardder** [comment]
Thanks for your amazing help!
[Here](https://ci.appveyor.com/api/buildjobs/413hfrcelrm2sd5v/artifacts/output%2Fnvda_snapshot_try-UIAScroll-17080%2C01fcd3d1.exe) is a try build that should fix that as well.

I'm actually pretty confident that I"ll be able to come up with a proper implementation for Internet Explorer. While it won't be as accurate as the other browsers, it will still allow us to have an implementation that is pretty consistent, which is more than I could have hoped before.

**Comment by k-kolev1985** [comment]
Yes, now the problem in Firefox is fixed. Thanks!

**Comment by leonardder** [comment]
Just to make sure, Edge and Firefox are behaving equally now?

**Comment by k-kolev1985** [comment]
> 
> 
> Just to make sure, Edge and Firefox are behaving equally now?

Yes, they do. :-)

**Comment by Adriani90** [comment]
@leonardder in case you are interested to work further on that pull request, I suggest to drop the implementation for the old Edge and stick to Edge Chromium instead.

**Comment by leonardder** [comment]
Note that the implementation for Edge Legacy/Spartan isn't very much Edge, but rather UIA specific. Therefore it also applies to Word with UIA enabled.

@JulienCochuyt are you still intending to adopt my work at some point? We really need a sighted person to work on this, I'm afraid. It's just too complex for blind eyes to be aware of all the edge cases.

**Comment by JulienCochuyt** [comment]
@leonardder wrote:
> @JulienCochuyt are you still intending to adopt my work at some point? We really need a sighted person to work on this, I'm afraid. It's just too complex for blind eyes to be aware of all the edge cases.

I really do. The big question is when will I have enough time to.
As you know, this issue is big enough not to be handled in a few minutes of my spare time, and will at least require a couple of days of my work time. As often, an emergency chases another, and despite this issue being in my work pipe, it couldn't yet gain a proper schedule.
If anyone feels qualified enough to pick it, I won't complain at all.
Otherwise, I'll do my best to prioritize it as much as can be.

**Comment by seanbudd** [comment]
@k-kolev1985 - is this bug still an issue in modern edge (based on chromium)?
